matching: Make `metavariable-type` default to false

IagoAbal opened the pull request:
Like every other metavariable-xyz operator.

In case we could not evaluate `metavariable-type`, we were defaulting to "true", that is, to let the range pass. But every other metavariable-xyz operator defaults to "false", if it cannot be evaluated then it filters out the range.

Also, this removes a `logger#info` reporting ranges filtered out by `metavariable-type` that was likely to be noisy, and no other `metavariable-xyz` operator logs such situations anyways.

Closes PA-3027

test plan:
```yaml
    patterns:
      - pattern: private int $X;
      - metavariable-type:
          metavariable: $Y # $Y is not bound
          type: int
```
no longer produces a match when run on
```java
class Test {
  private int x;
}
```

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

IagoAbal committed:
matching: Make `metavariable-type` to default to false

Like every other metavariable-xyz operator.

In case we could not evaluate `metavariable-type`, we were defaulting to
"true", that is, to let the range pass. But every other metavariable-xyz
operator defaults to "false", if it cannot be evaluated then it filters
out the range.

Closes PA-3027

test plan:
make test # new test

aryx commented on tests/rules/metavar_type_invalid.yaml:
```diff
@@ -0,0 +1,11 @@
+    patterns:
+      - pattern: private $T x;
+      - metavariable-type:
+          metavariable: $T
```
why this is not matching??? Or did you make a typo and you wanted to write 'metavariable: $X' (which is not bounded in the pattern, like you had in the changelog).

ihji replied:
> why this is not matching???

`$T` represents a type identifier so it doesn't match with a type itself. `private int $T` would match.

And I think Iago also wanted to write `metavariable: $X`.

aryx replied:
oh right, I got confused. the example is a bit confusing though.

IagoAbal replied:
Ooops, sorry, I wrote the wrong test indeed.

aryx commented on changelog.d/pa-3027.fixed:
````diff
@@ -0,0 +1,12 @@
+
+```yaml
+    patterns:
+      - pattern: private $T x
````
This example is confusing to me because I forgot metavariable-type is about checking the identifier
type, not the type itself.
So maybe `private int $X` and the metavariable-type: metavariable: $Y
would be better?
Even though we should really syntactically reject such rules in the metachecker.

IagoAbal replied:
Ignoring I wrote the wrong tests, I agree this should be caught by the meta checker. I was also thinking whether to write a test at all for this.

IagoAbal committed:
Yoann's review

IagoAbal committed:
Remove test

Cannot have a test that produces a matching error, even if that is just
actually a warning.

IagoAbal commented:
Once I fixed the test I actually cannot commit it because it makes the test suite fail. No "matching errors" allowed (despite they are more like warnings than fatal errors). So I just inlined the test example in the test plan.

IagoAbal commented:
> Exception raised: `missing field 'check_id' in JSON object of type 'CoreMatch'`

This error seems unrelated to this PR, probably Pro needs to be updated to use the latest Semgrep interfaces.

aryx commented:
I did update semgrep-pro, but we still get those errors.

IagoAbal committed:
Remove logging message

No other operator logs whether a range was filtered out, and that is
likely to be noisy.
